libraries/rush-lib/src/logic/operations/CacheableOperationPlugin.ts

```diff
@@ -0,0 +1,875 @@
+  // Controls the log for the cache subsystem
+  buildCacheTerminal: ITerminal | undefined;
+  buildCacheProjectLogWritable: ProjectLogWritable | undefined;
+}
+
+export class CacheableOperationPlugin implements IPhasedCommandPlugin {
+  private _buildCacheContextByOperationRunner: Map<IOperationRunner, IOperationBuildCacheContext> = new Map<
+    IOperationRunner,
+    IOperationBuildCacheContext
+  >();
+
+  public apply(hooks: PhasedCommandHooks): void {
+    hooks.createOperations.tapPromise(
+      PLUGIN_NAME,
+      async (operations: Set<Operation>, context: ICreateOperationsContext): Promise<Set<Operation>> => {
+        const { buildCacheConfiguration, isIncrementalBuildAllowed, cobuildConfiguration } = context;
+        if (!buildCacheConfiguration) {
+          return operations;
+        }
+
+        let disjointSet: DisjointSet<Operation> | undefined;
+        if (cobuildConfiguration?.cobuildEnabled) {
+          disjointSet = new DisjointSet<Operation>();
+        }
+
+        for (const operation of operations) {
+          disjointSet?.add(operation);
+          const { runner } = operation;
+          if (runner) {
+            const buildCacheContext: IOperationBuildCacheContext = {
+              // ShellOperationRunner supports cache writes by default.
+              isCacheWriteAllowed: true,
+              isCacheReadAllowed: isIncrementalBuildAllowed,
+              isSkipAllowed: isIncrementalBuildAllowed,
+              projectBuildCache: undefined,
+              cobuildLock: undefined,
+              cobuildClusterId: undefined,
+              buildCacheTerminal: undefined,
+              buildCacheProjectLogWritable: undefined
+            };
+            // Upstream runners may mutate the property of build cache context for downstream runners
+            this._buildCacheContextByOperationRunner.set(runner, buildCacheContext);
+
+            if (runner instanceof ShellOperationRunner) {
+              this._applyShellOperationRunner(runner, context);
+            }
+          }
+        }
+
+        if (disjointSet) {
+          // If disjoint set exists, connect build cache disabled project with its consumers
+          await Async.forEachAsync(
+            operations,
+            async (operation) => {
+              const { associatedProject: project, associatedPhase: phase } = operation;
+              if (project && phase && operation.runner instanceof ShellOperationRunner) {
+                const buildCacheEnabled: boolean = await this._tryGetProjectBuildEnabledAsync({
+                  buildCacheConfiguration,
+                  rushProject: project,
+                  commandName: phase.name
+                });
+                if (!buildCacheEnabled) {
+                  for (const consumer of operation.consumers) {
+                    if (consumer.runner instanceof ShellOperationRunner) {
+                      disjointSet?.union(operation, consumer);
+                    }
+                  }
```

Comment:
Provide code comments about the purpose of this exercise? In a normal build graph, it is not at all strange to have operations that use ShellOperationRunner have consumers that use NullOperationRunner, that in turn have consumers that use ShellOperationRunner, so the way this is computed is strange.

To use a very basic example, the various `heft-*-rig` projects don't have a build step, and therefore use `NullOperationRunner`, but they have dependencies that have build steps and consumers that have build steps.